divi/src/verifyDb.cpp

```diff
@@ -122,8 +122,25 @@ bool CVerifyDB::VerifyDB(const CCoinsView* coinsview, unsigned coinsTipCacheSize
             CBlock block;
             if (!ReadBlockFromDisk(block, pindex))
                 return error("VerifyDB() : *** ReadBlockFromDisk failed at %d, hash=%s", pindex->nHeight, pindex->GetBlockHash());
-            if (!ConnectBlock(block, state, pindex, coins, false))
+
+            /* ConnectBlock may modify some fields in pindex as the block's
+               status is updated.  In particular:
+
+                  nUndoPos, nStatus, nMint and nMoneySupply
+
+               In the current situation, we do not want to have the CBlockIndex
+               that is already part of the active chain modified.  Thus we
+               apply ConnectBlock to a temporary copy, and verify later on
+               that the fields computed match the ones we have already.  */
+            CBlockIndex indexCopy(*pindex);
+            if (!ConnectBlock(block, state, &indexCopy, coins, false))
                 return error("VerifyDB() : *** found unconnectable block at %d, hash=%s", pindex->nHeight, pindex->GetBlockHash());
+            if (indexCopy.nUndoPos != pindex->nUndoPos
+                  || indexCopy.nStatus != pindex->nStatus
+                  || indexCopy.nMint != pindex->nMint
+                  || indexCopy.nMoneySupply != pindex->nMoneySupply)
+                return error("%s: *** attached block index differs from stored data at %d, hash=%s",
+                             __func__, pindex->nHeight, pindex->GetBlockHash());
```

Comment:
ConnectBlock has some side-effects other than just connecting the block that you need to be be careful about. It allocates space for block data and block undo data to disk - it also writes to the lookup indices, as well as other things. I think ConnectBlock may need to be modified to accomodate for this:

Line 1029 in main.cpp should be something like:
    if(!fJustCheck && (!WriteUndoDataToDisk(pindex,state,blockTxChecker.getBlockUndoData()) ||
       !UpdateDBIndicesForNewBlock(indexDatabaseUpdates, chainstate.BlockTree(), state)))

Reply:
plus of course other changes to make it correctly update the coinsview

Reply:
That is a good point (even though it is not related to this PR).  The updates to the coinsview are intended even in this place, and `VerifyDB` uses a custom in-memory overlay view to make sure the checking works out (without modifying the real UTXO set).

The update to the undo data is only done if there is not yet an `nUndoPos`, so will be skipped here as these block indices already have it.  The index updates are idempotent, so while they will be done, it should not cause any harm.

However, I definitely agree that it would be better / easier to understand and verify if those updates were not done at all instead.  The `fJustCheck` flag already does this actually (by an early return before the line you quote), but is currently never set to `true` (also not for `VerifyDB`).  It would need to be modified to be useful, as we still want the `view.SetBestBlock` call to be made also for the `VerifyDB` flow.  I think that's an interesing suggestion, though, I will prepare an update.

Reply:
This has been implemented in #114 now, which we should review & merge first, and then revisit the current PR.

Reply:
#114 is merged, and I've rebased this on top.